File: Data Structures & Algorithms/depth-of-binary-tree/submission-0.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def maxDepth(self, root: Optional[TreeNode]) -> int:

        def dfs(node, depth):
            if not node:
                return depth

            depth += 1

            depth_left = dfs(node.left, depth)
            depth_right = dfs(node.right, depth)

            depth -= 1

            return max(depth_left, depth_right)
        
        return dfs(root, 0)

File: Data Structures & Algorithms/binary-tree-diameter/submission-0.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
private:
    int ans = 0;

    int dfs(TreeNode* node) {
        if (!node) return 0;

        int leftDepth = dfs(node->left);
        int rightDepth = dfs(node->right);

        ans = max(ans, leftDepth + rightDepth);

        return max(leftDepth, rightDepth) + 1;
    }

public:
    int diameterOfBinaryTree(TreeNode* root) {
        dfs(root);
        return ans;
    }
};
